### PR TITLE
feat(api): Jwt

### PR DESCRIPTION
**Pull request by JeanVittory**
## Summary

The module auth was completed for the user. Currently, it's possible to be registered, authorized and authenticated. 

### TO-DO

So, it's necessary to build the registration process of doctors to complete the auth module

### CODE REVIEW

File: src/api/schemas/authentication.schema.ts [hunk 1]
```diff
@@ -1,8 +1,8 @@
 import Joi from 'joi';
 
-const loginSchema = Joi.object({
+const autenticationSchema = Joi.object({
```
Comment (ht1204) [inline src/api/schemas/authentication.schema.ts]: `authenticationScheme`

File: src/api/service/Auth.service.ts [hunk 2]
```diff
@@ -14,50 +21,47 @@ const prisma = new PrismaClient();
 export class AuthService {
   constructor() {}
   //TO-DO: finish the doctors and admin login in order to correctly take the output of the login method.
```
Comment (ht1204) [inline src/api/service/Auth.service.ts:25]: is this To-Do done?

File: src/api/service/Auth.service.ts [hunk 3]
```diff
@@ -14,50 +21,47 @@ const prisma = new PrismaClient();
...
+      return token;
     } catch (error: any) {
       throw error;
     }
   }
 
-  private static async getRole(role: number): Promise<{ id: string }> {
+  private static signToken(id: string): IAuthentication {
+    const ACCESS_TOKEN: string = JWT.sign({ id }, env.SECRET_JWT, {
+      expiresIn: Math.floor(Date.now() / 1000) + 60 * 60,
```
Comment (ht1204) [inline src/api/service/Auth.service.ts]: This value: `Math.floor(Date.now() / 1000) + 60 * 60` could be better in a constant

File: src/api/service/Auth.service.ts [hunk 4]
```diff
@@ -14,50 +21,47 @@ const prisma = new PrismaClient();
...
+    });
+    return { ACCESS_TOKEN: ACCESS_TOKEN };
```
Comment (ht1204) [inline src/api/service/Auth.service.ts]:
```js
return {
  ACCESS_TOKEN
};
```

File: src/api/service/Auth.service.ts [hunk 5]
```diff
@@ -76,4 +80,19 @@ export class AuthService {
       }
     }
   }
+
+  private static async getRole(role: number): Promise<{ id: string }> {
+    try {
+      return (await prisma.role.findFirst({
+        where: {
+          role: role,
+        },
```
Comment (ht1204) [inline src/api/service/Auth.service.ts:101]:
```js
return (await prisma.role.findFirst({
        where: {
          role,
        },
```

File: src/api/service/Auth.service.ts [hunk 6]
```diff
@@ -76,4 +80,19 @@ export class AuthService {
...
+        select: {
+          id: true,
+        },
+      })) as { id: string };
```
Comment (ht1204) [inline src/api/service/Auth.service.ts]: `{ id: string }` there could be a custom datatype for this.

File: src/api/service/Auth.service.ts [hunk 7]
```diff
@@ -76,4 +80,19 @@ export class AuthService {
...
+    } catch (e) {
+      throw e;
```
Comment (ht1204) [inline src/api/service/Auth.service.ts]: meaningful names: `err: Error` or `error:  Error`

Comment (ht1204) [reply src/api/service/Auth.service.ts]: So, I would like you to consider [this](https://kentcdodds.com/blog/get-a-catch-block-error-message-with-typescript).

File: src/api/service/Auth.service.ts [hunk 8]
```diff
@@ -14,50 +21,47 @@ const prisma = new PrismaClient();
...
-          id: true,
-        },
-      })) as { id: string };
-    } catch (e) {
-      throw e;
+      const user = await Users.getUser(id);
+      const userFormatted: IUserResponse = new UserDTO(user);
+      return userFormatted;
+    } catch (error) {
+      if (error instanceof Prisma.PrismaClientKnownRequestError) {
```
Comment (ht1204) [inline src/api/service/Auth.service.ts:67]: Excellent.

File: src/api/service/Auth.service.ts [hunk 9]
```diff
@@ -14,50 +21,47 @@ const prisma = new PrismaClient();
...
-      } = response;
-      const user: IUserResponse = { ...userData, role };
-      return user;
+      const response = await Users.getUser(email);
+      if (!response) return ApiError.Unauthorized();
+      const isAuth = await bcrypt.compare(password, response.password);
+      if (!isAuth) return ApiError.Unauthorized();
+      const token = this.signToken(response.id);
+      return token;
     } catch (error: any) {
```
Comment (ht1204) [inline src/api/service/Auth.service.ts]: Please, read [this](https://kentcdodds.com/blog/get-a-catch-block-error-message-with-typescript)

**Review by ht1204 — CHANGES_REQUESTED** [review]
You have done a great job, but please, consider and address the comments for possible enhancements.

File: src/api/service/Auth.service.ts (hunk 2 above)
Comment (JeanVittory) [reply src/api/service/Auth.service.ts:25]: no, it has not been done yet. keep being a TO-DO

File: src/api/service/Auth.service.ts (hunk 4 above)
Comment (JeanVittory) [reply src/api/service/Auth.service.ts]: done!

File: src/api/service/Auth.service.ts (hunk 5 above)
Comment (JeanVittory) [reply src/api/service/Auth.service.ts:101]: done!

File: src/api/service/Auth.service.ts (hunk 7 above)
Comment (JeanVittory) [reply src/api/service/Auth.service.ts]: thank you for this! i did something like that but using the prisma errors directly:
```
      if (error instanceof Prisma.PrismaClientKnownRequestError) {
        logger.error(error);
        logger.info('Prisma error:', error);
        if (prismaErrorsCodes400.includes(error.code)) throw new PrismaError(error.message, 400);
        if (prismaErrorsCodes404.includes(error.code)) throw new PrismaError(error.message, 404);
        throw new PrismaError(error.message, 500);
      }
```
in every method directly called by the controller i managed the error with the  validation before mentioned... In your suggestion the method getRole or userAuthentication are called from their respective services and there you will find the handle error. I would prefer just throw the error from the getRole or userAthentication and manage the error directly from the service an then at controller return the respective response

File: src/api/service/Auth.service.ts (hunk 9 above)
Comment (JeanVittory) [reply src/api/service/Auth.service.ts]: as i mentioned before this is being managed in the method that called this function. Please check my last commit for be updated about this topic

File: src/api/service/Auth.service.ts (hunk 2 above)
Comment (ht1204) [reply src/api/service/Auth.service.ts:25]: ok, thx 4 clarifying.

File: src/api/service/Auth.service.ts (hunk 7 above)
Comment (ht1204) [reply src/api/service/Auth.service.ts]: that was a great implementation and I considered to complement as well, do not forget the error management every time as you can.

**Review by ht1204 — APPROVED** [review]
Great Job.

**Pull request merged**